libraries/AP_GroundEffectControl/AP_GroundEffectController.cpp

```diff
@@ -0,0 +1,189 @@
+    // @User: Standard
+
+    // @Param: D
+    // @DisplayName: D gain
+    // @Description: D gain
+    // @User: Standard
+
+    // @Param: IMAX
+    // @DisplayName: IMax
+    // @Description: Maximum integrator value
+    // @User: Standard
+    AP_SUBGROUPINFO(_pitch_pid, "_PITCH_", 3, GroundEffectController, PID),
+
+	// @Param: THR_REF
+	// @DisplayName: Ground Effect desired throttle (percentage)
+	// @Description:
+	// @Range: 0.0 100.0
+	// @Increment: 0.01
+    // @User: Standard
+	AP_GROUPINFO("_THR_REF",   4, GroundEffectController, _THR_REF,   35.0),
+
+	// @Param: THR_MIN
+	// @DisplayName: Ground Effect minimum throttle (percentage)
+	// @Description:
+	// @Range: 0.0 100.0
+	// @Increment: 0.01
+    // @User: Standard
+	AP_GROUPINFO("_THR_MIN",   5, GroundEffectController, _THR_MIN,   20.0),
+
+    // @Param: THR_MAX
+	// @DisplayName: Ground Effect maximum throttle (percentage)
+	// @Description:
+	// @Range: 0.0 100.0
+	// @Increment: 0.01
+    // @User: Standard
+	AP_GROUPINFO("_THR_MAX",   6, GroundEffectController, _THR_MAX,   50.0),
+
+	// @Param: ALT_REF
+	// @DisplayName: Ground Effect desired altitude (meters)
+	// @Description:
+	// @Range: 0.0 1.0
+	// @Increment: 0.01
+    // @User: Standard
+	AP_GROUPINFO("_ALT_REF",   7, GroundEffectController, _ALT_REF,   0.45),
+
+	// @Param: CUTOFF_FREQ
+	// @DisplayName: Rangefinder Complementary Filter Cutoff Frequency
+	// @Description: Lower values will trust the rangefinder less.
+	// @Range: 0.0 2.0
+	// @Increment: 0.01
+    // @User: Advanced
+	AP_GROUPINFO("_CUTOFF_FRQ",   8, GroundEffectController, _CUTOFF_FREQ,   0.1),
+
+	// @Param: LIM_ROLL
+	// @DisplayName: Max roll angle (degrees)
+	// @Description: Max roll allowed in auto mode while going towards a ground effect waypoint. 0 to disable.
+	// @Range: 0.0 45.0
+	// @Increment: 0.01
+    // @User: Advanced
+	AP_GROUPINFO("_LIM_ROLL",   9, GroundEffectController, _LIM_ROLL,   10.0),
+
+    AP_GROUPEND
+};
+
+bool GroundEffectController::user_request_enable(bool enable)
+{
+    if(enable){
+        if(!_ACTIVE || !_rangefinder->has_orientation(ROTATION_PITCH_270)){
+            _enabled = false;
+            return false;
+        }
+    }
+
+    _enabled = enable;
+    return true;
+}
+
+void GroundEffectController::reset()
+{
+    _altFilter.set_cutoff_frequency(_CUTOFF_FREQ);
+    _altFilter.reset();
+
+    _pitch_pid.reset_I();
+    _throttle_pid.reset_I();
+    return;
+}
+
+int32_t GroundEffectController::get_auto_lim_roll_cd()
+{
+    if(_LIM_ROLL <= 0.0001f){
+        return INT32_MAX;
+    }
+    return int32_t(_LIM_ROLL*100.0);
+}
+
+void GroundEffectController::update()
+{
+    uint32_t time = AP_HAL::micros();
+    if(time - _last_time_called > RESET_TIMEOUT_MICROS){
+        reset();
+    }
+    _last_time_called = time;
+
+    if(_rangefinder->status_orient(ROTATION_PITCH_270) == RangeFinder::Status::Good) {
+        _last_good_rangefinder_reading = _rangefinder->distance_orient(ROTATION_PITCH_270);
```

Comment:
I would multiply by sin(roll angle). This will make the height track more stable in turns. 

_last_good_rangefinder_reading = sin(roll_angle) * (_rangefinder->distance_orient(ROTATION_PITCH_270));

Reply:
This depends if the rangefinder measures:
a) the closest object in its field-of-view, or
b) the straight-line distance to where it is pointing

If the answer is a, I think that cos() is not necessary. The user just needs to avoid banking more than the FOV/2. Does anybody know if downward-facing rangefinders in copter are rotated like this?

Roll is already a problem because the dynamics change and because of wing strikes. I added a roll-limit parameter to avoid all of these issues for now.

Reply:
This is what copter does, https://github.com/ArduPilot/ardupilot/blob/063e24c7239484b4ddced0da40a80f575d5c8e61/ArduCopter/sensors.cpp#L34

It takes into account both roll and pitch.